Add e2e tests with Cypress

trumbitta opened the pull request:
## Description

🌲 This adds Cypress and some first e2e tests (let's talk about how many we want to add).

🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊
**IMPORTANT! See TODO below!** 
🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊

### 🤔 Reasoning

Apart from the obvious "e2e tests - especially with Cypress - are good for any web application" thing, I wanted to add some tests before tackling the warnings & C. [@gtsiolis and I discussed about on Discord](https://discord.com/channels/816244985187008514/914836798830690344/930546351698247690).

🍩  I'd feel way more comfortable in changing some of the things that need to be changed if I had some tests helping me find any obvious problems deriving from my fixes.

### 📣 Current status

So far I managed to install Cypress, have it run both headless (for CI) and headed (for debugging and implementation of test cases), and I started writing the first test cases.

I also activated TypeScript support and installed Cypress Testing Library (CTL).

You may also find out I updated a `div` to `nav` because it was the semantic thing to do, and because it helped me write a better test case with CTL.

And I updated the service mock with the simplest implementation of two missing methods.

### Updated TODO

- [x] Rebase from main, update test spec (See https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1014470441)
- [x] Fix conflicts: I know how to deal with `App.tsx`, I'll need some advice @geropl for the ones specifying the docker image (See https://github.com/gitpod-io/gitpod/pull/7611#discussion_r788202086)
- [x] Add quick and easy instructions about tests to README (See again https://github.com/gitpod-io/gitpod/pull/7611#discussion_r788202086 for an interesting gotcha)
- [x] Restore `craco.config.js`
- [x] Reorganize commits (interactive rebase to properly group commits together)

### ⚠️ Original TODO

- Decide where to stop with this first step (MVC)
- Write more tests
- Ask for guidance about properly updating the docker image and the structural changes at large (I basically hacked together things by @mikenikles in a way that's temporary to say the least)
- Find a way to test the unauthenticated scenario (log in page etc)
- Find a way to handle `craco.config.js` so that when running in CI you don't have to customize it (Nx has environments and environment bound configurations, but at this time I don't yet know about craco / cra)
- Update dashboard README with some info about how to write, run, and maintain e2e tests
- Polish

## How to test
<!-- Provide steps to test this PR -->

- go inside the dashboard directory: `cd components/dashboard`
- in one terminal, launch the dashboard app: `yarn start`
- in another terminal, launch the tests: `yarn cypress install && yarn test:e2e` (re-installation of cypress is something I have to get rid of)
- open port `6080` in the browser to see Cypress via VNC

| Cypress via VNC |
|---|
|![image](https://user-images.githubusercontent.com/462705/149536308-9f50e624-57d0-4a3b-ad4a-5f8b88b7eb22.png)|

- find and click on `app.spec.ts` (the only spec file **at the time of writing**)
- tests should run and pass


## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **alextugarev** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @alextugarev` in a comment when ready.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["alextugarev"]} -->

aledbf commented on .gitpod.Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    emacs-gtk
+
+# Install Cypress dependencies.
+RUN sudo apt-get update \
```
Please use the same approach `sudo install-packages`

trumbitta replied:
@aledbf thanks for taking the time to look at this 🙏 

🥷🏻 I took this part directly from https://github.com/mikenikles/cypress-on-gitpod/blob/master/.gitpod.Dockerfile but I don't expect it to stay like this.

🔮 I'm guessing that if the PR ends up getting accepted, all the packages and configurations in this file should be directly included 📦 in the docker image (e.g. `eu.gcr.io/gitpod-core-dev/dev/dev-environment:cw-bump-leeway.29`) instead.

🎯 This is the reason why I have a bullet point in the TODO (see the description of this PR) about getting someone from Gitpod to help me better understand, and implement, this update of the usual docker image:

> Ask for guidance about properly updating the docker image and the structural changes at large (I basically hacked together things by @mikenikles in a way that's temporary to say the least)

csweichel commented:
Thanks @trumbitta for contributing this.

If the @gitpod-io/engineering-meta wants to pull this in, we'll want to find a way to integrate this with our existing integration tests (not necessarily before merging this PR), but short-term.

trumbitta commented:
100% @csweichel and let me know if I can help with that.

📢 Update:
I just added some more test bits, and in doing so I also refactored into constants some magic strings that were used for route paths.

🙈 I also noticed that with https://github.com/gitpod-io/gitpod/commit/932a33a414e2afb20f9e4acf66664cd38356c7b8#diff-9ace031ab25e089e2961e66b81b91b60cbc8d6043b2f3cc03827f68c9886f223 "Projects" is not the starting page anymore and instead "Workspaces" is, making the very first test already fail 😄 

👻 All of the above resulted in @gtsiolis appear in the night sky as a [Force ghost](https://starwars.fandom.com/wiki/Force_ghost) (but alive and well) whispering "MVC! MVC!" at me.

💪 This is why, if @gitpod-io/engineering-meta thinks this PR is a good idea, I think I should stop writing more tests and we should proceed in the following general direction:

- somebody from Gitpod helps streamline and polish the new docker confs I hacked together 🧙 
- I rebase from main, fix the easy (🥳 ) conflict, and update the only test to the new situation ⛑️ 
- review + merge 🍻 
- we start thinking about the next steps / tests cases 🔮

geropl commented:
Hi @trumbitta , thank you for all the effort you put into this! :muscle:
I will now start reviewing, and get back to you tomorrow morning (latest).

geropl commented on components/dashboard/package.json:
```diff
@@ -48,6 +50,8 @@
     "start": "craco start",
     "build": "craco build --verbose",
     "test": "craco test",
+    "test:unit": "craco test",
```
Consequently, I'd remove the `test:unit` target here. Or rename it to `test:interactive`/`test:watch`, maybe...?

trumbitta replied:
See reply under https://github.com/gitpod-io/gitpod/pull/7611#discussion_r786737456

geropl commented on components/dashboard/package.json:
```diff
@@ -48,6 +50,8 @@
     "build": "craco build --verbose",
     "test": "craco test",
+    "test:unit": "craco test",
+    "test:e2e": "cypress open",
```
For me this is not about e2e testing at this point. I'd rather rename this to `cypress` or so, do indicate that this opens the interactive test runner, which might be very helpful for testing.

trumbitta replied:
I added `test:unit` and `test:e2e` as variations since the first one uses Jest for the couple unit tests that are already there (by @laushinka IIRC), while the second one is the new Cypress one.
I wanted to delete the now redundant default "test" script created by Create React App but I didn't want to accidentally break any CI job that might be using it to launch those couple unit tests.

## My proposal

- Delete the "test" script and hope for the best 🤞 about CI (joking 😄 do you know of any such job that might need updating to point at the new "test:unit" script?)
- For unit tests, Craco and CRA run jest in watch mode by default. To avoid that, CI can set a `CI=true` env variable or add `--watchAll=false` to the command line 👌 
- Rename "test:e2e" into "test:integration:watch"
- Add new "test:integration:run" script (`cypress run`) aimed at CI and automation

geropl replied:
The "test" command can be configured here: https://github.com/gitpod-io/gitpod/blob/28bf0e6541255d43ea1a317ff852ee79b9e14147/components/dashboard/BUILD.yaml#L20-L21
That should be set to `test: ["yarn", "test"]`.

`yarn test` by convention (see other TS components) is the script that executes all "tests that are executed on parts of the code base, do not have component-external dependencies, are non-interactive and have quick turn arounds". This is to enable everyone - without deeper knowledge of the component, or tech involved - to run these _the same way CI does_. I think this is a very nice property worth maintaining. We can discuss for sure, but that's maybe out of scope of the PR.

All other scripts are there to offer more fine grained control; :+1: for your suggestions!

trumbitta replied:
✅ scripts updated

I didn't add the test command to BUILD.yaml though, because I noticed `dontTest: true` and I'm not sure what it means. I guess it can be done in a different PR?

geropl replied:
>  didn't add the test command to BUILD.yaml though, because I noticed dontTest: true and I'm not sure what it means. I guess it can be done in a different PR?

:100:

geropl reviewed (commented):
I really really like this! I see at least two different levels where this is useful:

## Unit test
In some sense we've (mis-)treated `dashboard as "not being real code" so far, which explains why we do not have any tests for it. On the other hand, it grew quite complex in some corners, and will continue to do so all over the place to enhance the UX.
Being able to express (selected) invariants as code/tests improves our confidence in the changes we make, and helps us immensely to live up to the expectations in stability our users have.

Using a mock service here is a tradeoff - and limited -, of course, which we should be conscious about: The tests we add here should be about rendering and UX only, not about how some flows behave. Still I see a lot of potential value; some good examples include:
 - does startWorkspace screen has button "Skip prebuild"
 - what's the default page for 1st time users
 - what's the default page for Gitpod users
 - test button interaction in complex modals

I think the state of the PR (sans some polishment, see Dockerfile and package.json) is the perfect starting point, and already helps with that from day 1 (the tests already implemented).

## Integration/End-to-end tests
The frontend interaction part with cypress could help a ton in testing Gitpod in general: With minimal effort one could test a lot of things across the complete stack. This comes with the well-known the downsides of not being very precise, potentially very flaky and hard to execute right.

While cypress solves the frontend interaction part, IMO we should treat this as a second, separate step and use case because:
 - unit tests have their own merrit
 - setting up and executing end-to-end tests reliably is a far bigger task that we lack a lot of other things for as well, and is far beyond the scope of this PR
 - we might lack the resources to pull this off right now

## How to move on
@trumbitta Either you try and go ahead with the suggested changes, or one of us does the Dockerfile merge/integration and other comments.
Once that's done we can merge! :rocket:

geropl commented on .gitpod.Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    xvfb \
+    && sudo rm -rf /var/lib/apt/lists/*
+
+# Overwrite this env variable to use a different window manager
```
E.g, [this](https://github.com/gitpod-io/gitpod/blob/main/dev/image/Dockerfile#L5) should be `gitpod/workspace-full-vnc:latest`. Then all scripts can be removed and their paths adapted.

trumbitta replied:
🤞 Ok, I think I fixed the docker conf in 5619a174299759e47904f14969ca0a61e70c7049 but I don't know how to launch a new build for the image so that a new release appears [on GCR](https://console.cloud.google.com/gcr/images/gitpod-core-dev/eu/dev/dev-environment) and I can then replace the original `:cw-bump-leeway.29` 🐳  and have a working environment again.

Any pointers? 😊

I need this to work in order to update the test that will fail after the rebase from `main` as [mentioned in a previous comment](https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1014470441).

geropl replied:
Whoops, answer went here: https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1016327106

trumbitta replied:
Thanks, @geropl, it worked!

A couple of problems though:

- all the files where the image is specified are now conflicting
- I found out this hacky bit from the `workspace-full-vnc` image:
```sh
# This is a bit of a hack. At the moment we have no means of starting background
# tasks from a Dockerfile. This workaround checks, on each bashrc eval, if the X
# server is running on screen 0, and if not starts Xvfb, x11vnc and novnc.
RUN echo "export DISPLAY=:0" >> ~/.bashrc
RUN echo "[ ! -e /tmp/.X0-lock ] && (/usr/bin/start-vnc-session.sh &> /tmp/display-\${DISPLAY}.log)" >> ~/.bashrc
```
makes Cypress work only if you are using bash, or you load bash then switch back to your favorite shell. For example my default shell is zsh.
I'll add a note in the README for now, but it feels like something that should be addressed sooner or later.

geropl replied:
> all the files where the image is specified are now conflicting

This is a general problem with how we manage this dev image.  Conflicts are just very likely to happen. :grimacing: 

@trumbitta I propose the following: you sort out the other todos, incl. rebase. Then ping me for another build, and we put the resulting image on top and merge quickly. Does that sound good?

> makes Cypress work only if you are using bash, or you load bash then switch back to your favorite shell. For example my default shell is zsh.

Agree this is hacky. I'd leave it for now, and tackle that in a separate PR. Not sure, maybe `/etc/profile` is a better place for this.

trumbitta replied:
> @trumbitta I propose the following: you sort out the other todos, incl. rebase. Then ping me for another build, and we put the resulting image on top and merge quickly. Does that sound good?
> 

🤝 perfect

trumbitta replied:
@geropl I think we are ready for the new build!

jankeromnes commented on components/dashboard/src/App.tsx:
```diff
@@ -261,32 +264,32 @@ function App() {
-                <Route path="/variables" exact component={EnvironmentVariables} />
-                <Route path="/preferences" exact component={Preferences} />
-                <Route path="/install-github-app" exact component={InstallGitHubApp} />
+                <Route path={workspacesPathMain} exact component={Workspaces} />
```
I'm not sure what the advantage is of replacing a simple `<Route path="/literal-path">` with something more complicated like `<Route path={somethingDynamicThatMightFailOneDay}>`. 😬

In the current code, moving, renaming, or adding a new page is pretty simple (requires a single line change).

In the new proposed code, you need to export a new variable somewhere, import it in this file, and use it here (requires a coordinated multi-file change, which is more complicated for the same value).

Maybe I'm missing an important use-case that the more complicated routing allows? (Sorry, haven't read the entire Pull Request yet)

trumbitta replied:
@geropl Years ago I used to always have something like `routes.config.ts` with all the routes of the application in a single file. An approach I had borrowed from early releases of Sails.js and it made perfect sense. Like a command center for routes.
But then 4-ish years ago [Nx](https://nx.dev/) changed [how I envision and tend to organise an application](https://egghead.io/courses/scale-react-development-with-nx-4038):
- We can think of each top-level route and component as the head of something we can call "feature" (workspaces, projects, settings, etc are all "features")
- Some of the folders you Gitpodders already organised the code into, tell us you already were thinking in such terms: we already have a folder for "settings", a folder for "workspaces", ...
- Now, what if these features were also units, black boxes, something that works sometimes depending on other units but substantially and primarily by itself. After all, when you visit `/settings` you mostly (only?) do things that are provided by the components in the `settings` folder + some other shared utility components. Nx calls those units "libs" and lets you organise them into actual libs you can lint, test, build, reason about separately... but I digress.
- Given all of the above, having the route paths for a feature in a file inside the feature folder starts making more sense because everything that might happen to those files - moving, renaming, editing, whatever - it will always happen in the same place: the feature folder. And "where are the routes for this feature? Oh right, in the feature folder, of course".
In a word: colocation.

@jankeromnes I'm all for simple magic strings when they are used in just one place.
With the new tests needing the paths to visit and check, the situation changes though.
Having the paths stored in configuration constants:
- enables safe reuse between components and spec files (something we need today)
- enables safe reuse throughout the application wherever you may need to redirect the user from one page to another programmatically or by providing a link:
"ok in this new page X I need to provide a link to the workspaces page... what was the path again? Oh no! I need to check somewhere and find out!" vs "ok in this new page X I need to provide a link to the workspaces page... what was the path again? Easy, I know the convention: feature name, then it's a path, and the main one: `workspacesPathMain` oh, thank you IntelliSense and auto-imports! 🚀"
- protects them from accidental edits (the component files are subject to more frequent changes than a simple configuration file)
- enables robust combination. E.g.:
```ts
export const settingsPathTeams = '/teams';
export const settingsPathTeamsJoin = [settingsPathTeams, 'join'].join('/');
export const settingsPathTeamsNew = [settingsPathTeams, 'new'].join('/');
```
Where you can't accidentally mistype or edit the second and third `/teams` because they literally don't exist, and where what really counts ("join" and "new") is immediately visible
- has the only cost of exporting a const from a file, then you just use it wherever you need it (not just one single file anymore - since we now have test that are using them) by also leveraging the auto-import feature of your favourite IDE

geropl replied:
> I'm all for simple magic strings when they are used in just one place.
With the new tests needing the paths to visit and check, the situation changes though.

Makes perfect sense to me. :+1:

trumbitta commented:
> ## How to move on
> @trumbitta Either you try and go ahead with the suggested changes, or one of us does the Dockerfile merge/integration and other comments. Once that's done we can merge! 🚀

🎉  Hey @geropl, thanks for the review and happy you find this useful!
I 100% agree with your premise (not quoted here).

I'll try doing the 💅 polishing on my own re: docker 🐳 (thanks for the directions!), and I'll reply to / act on the comments as soon as I am able 🛠️ 

🗒️ Addendums:
- I didn't actually add anything about unit tests; I think I saw a couple tests by @laushinka while I was scouting around the codebase before starting with this
- Re: the dashboard "not being real code".
Since the first time I started looking into the dashboard, I recognized the typical signs of code written by a skilled, but overwhelmed, team of mostly full-stack devs in a successful, fast-growing, startup.
That's why I decided to start proposing my somewhat "heavy" PRs / ideas: I thought "ok, as a seasoned frontend specialist, I can help. We can consolidate this, make it more robust, organized, a solid foundation to support the even faster growth that's likely to be coming soon."
And this also gets me to the next point:
- Re: "we don't have the resources".
👋 Hey, it's me: I'm resources 😄 at least to the extent of my free time and willpower 🤞

geropl commented:
> I didn't actually add anything about unit tests

@trumbitta I might have mis-used the term a bit, let me explain:
When referring to "unit tests" I actually meant "tests that are executed on parts of the code base, do not have component-external dependencies, are non-interactive and have quick turn arounds". Those are the ones we execute with `yarn test` by convention We sometimes (falsely) refer to those as "unit test". But from the point of the dashboard component it's actually a "integration test", I agree :100: . Will adjust my wording.

trumbitta commented on components/dashboard/craco.config.js:
```diff
@@ -5,12 +5,21 @@
+      plugins: [require("tailwindcss"), require("autoprefixer")],
+    },
+  },
+  devServer: {
```
Note to self: remember to delete this part after everything else is ok

trumbitta commented on scripts/novnc-index.html:
```diff
@@ -0,0 +1,8 @@
+<!--
```
Yep, see https://github.com/gitpod-io/gitpod/pull/7611#discussion_r787174895

trumbitta commented on scripts/start-vnc-session.sh:
```diff
@@ -0,0 +1,15 @@
+# see https://en.wikipedia.org/wiki/Xvfb#Remote_control_over_SSH
+
+DISP=${DISPLAY:1}
+
```
See https://github.com/gitpod-io/gitpod/pull/7611#discussion_r787174895

trumbitta commented on components/dashboard/src/projects/projects.routes.ts:
```diff
@@ -0,0 +1,11 @@
+/**
```
TL;DR: nobody ever, in any possible situation or personal condition ever, will have to think twice about which feature this file refers to (projects) or what it contains (routes).

This is something I started doing when I realised that in some contexts we don't usually think about, a file could be misinterpreted or colleagues I was sharing it with weren't immediately sure of which file it was and to which mental context they needed to switch in order to have a conversation about it. Whereas to me it was obvious what it was because I had just picked it and shared it.
Copy it around, or share it, or move it, or open it where its full path doesn't fit and you can only read its name, or ask a sleep deprived colleague to take a look, etc, etc.
Nobody ever will be unsure about what a file with a name like this is.
It lightens the cognitive load for developers.
CC @loujaybee because this is an oft neglected example of DevX 😉

geropl replied:
:+1:

trumbitta commented on components/dashboard/package.json:
```diff
@@ -48,6 +50,8 @@
     "start": "craco start",
     "build": "craco build --verbose",
     "test": "craco test",
```
See reply under https://github.com/gitpod-io/gitpod/pull/7611#discussion_r786737456

geropl commented:
/werft run

geropl commented:
@trumbitta I triggered a build job: https://werft.gitpod-dev.com/job/gitpod-build-dashboard-e2e-fork.0
It's not done yet, but the `dev` image is already there: `eu.gcr.io/gitpod-core-dev/dev/dev-environment:dashboard-e2e-fork.0`

To update simply search&replace across the whole code base and commit that on top.

trumbitta commented:
Ok I think we are getting close.
See [the updated TODO in the description](https://github.com/gitpod-io/gitpod/pull/7611#:~:text=two%20missing%20methods.-,Updated%20TODO,-Rebase%20from%20main).

Going to sleep now, will get back at it ASAP.

geropl commented:
/werft run

trumbitta commented:
@geropl I updated the image in the various files and I'm now doing some final checks 🤞 
Will report back!

trumbitta commented:
> @geropl I updated the image in the various files and I'm now doing some final checks 🤞 Will report back!

✅  @geropl tests work, the dashboard app works 🚀

geropl commented:
@trumbitta If you mark the PR "ready ro review" I'm happy to :heavy_check_mark:

geropl commented:
/werft run

geropl commented:
/lgtm

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 633256c1318b3e6d71eadb62b64e0f33e117abb0</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1017568789" title="LGTM">geropl</a>*

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [geropl]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1017569143" title="Approved">geropl</a>*

Associated issue requirement bypassed by: *<a href="https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1017569143" title="Approved">geropl</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [geropl]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

trumbitta commented:
Thanks everyone! It was a pleasant and satisfying experience from start to finish! 🧡
